rllib/evaluation/env_runner_v2.py

```diff
@@ -785,12 +785,13 @@ def _handle_done_episode(
             is_error = True
             episode_or_exception: Exception = env_obs_or_exception
         else:
-            is_error = False
-            # Output the collected episode.
-            self._build_done_episode(env_id, is_done, hit_horizon, outputs)
             episode_or_exception: EpisodeV2 = self._active_episodes[env_id]
             # Add rollout metrics.
             outputs.extend(self._get_rollout_metrics(episode_or_exception))
+            is_error = False
+            # Output the collected episode after (1) rollout metrics so that we
+            # always fetch metrics with RolloutWorker before we fetch samples
+            self._build_done_episode(env_id, is_done, hit_horizon, outputs)
```

Comment:
No but it appends to outputs! And because of the yield pattern, we now yield the metrics before we yield the batch that comes with it. This makes it so that we are sure to have yielded the metrics!

Reply:
as we discussed offline, hopefully we don't have any logic (in code, or in tests) that depend on the order of the data bits in output queue. that's just not part of the spec.
thanks.

Reply:
test_rollout_worker.test_metrics depends on the metrics being returned with/before all SampleBatcheType items.

Reply:
In a second offline discussion we decided to go with this behaviour here, because that's also what env runner "v1" does.

Reply:
yeah this is bad design that we are inheriting for now. The order should not matter. We can roll with this now, if @gjoliver does not have any graceful short-term suggestions here.

Reply:
something easy we could do here is not modify outputs in place, or if we are going to, still have it return outputs and set it to itself, so that its easier to understand that outputs is being modified

Reply:
Imho we should always yield outputs as the whole list so that we always yield metrics and last batch with the same call. @avnishn @kouroshHakha @gjoliver

Reply:
or just don't yield different things in a single queue ... :)